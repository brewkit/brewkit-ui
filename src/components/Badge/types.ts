import { ReactNode } from 'react';


export type Anchors = 'top-right' | 'top-left' | 'bottom-left' | 'bottom-right';
export type Colors = 'success' | 'warning' | 'danger' | 'info';


export interface Props {

    /**
     * Where the badge is positioned
     */
    anchorFrom?: Anchors,

    /**
     * The child element you want to apply the badge to
     */
    children?: ReactNode,

    /**
     * Changes badge style to circular
     */
    isCircular?: boolean,

    /**
     * Additional custom class to attach
     */
    className?: string,

    /**
     *
     */
    content?: string,

    /**
     * Color of the badge
     */
    color?: Colors,

}
