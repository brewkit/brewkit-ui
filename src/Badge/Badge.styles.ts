import { createStyles, makeStyles } from '@material-ui/core';

export default makeStyles(({ palette }) =>
    createStyles({
        badge: {
            borderRadius: '50%',
        },

        colorPrimary: {
            color: palette.common.white,
        },

        colorSecondary: {
            color: palette.common.white,
        },

        colorError: {
            color: palette.common.white,
        },
    }),
);
