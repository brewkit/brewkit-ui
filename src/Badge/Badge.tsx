import * as React from 'react';
import { merge } from 'lodash';
import MUIBadge, { BadgeProps as MUIBadgeProps } from '@material-ui/core/Badge';
import useStyles from './Badge.styles';

export type BkBadgeProps = MUIBadgeProps;

/**
 * Badge generates a small badge to the top-right of its child(ren).
 *
 * [Material-UI Docs](https://material-ui.com/components/badges/)
 *
 * > No differences in Brewkit.
 */
export const Badge = React.forwardRef(
    (props: BkBadgeProps, ref: React.Ref<any>): React.ReactElement => {
        const { classes: userClasses = {}, color = 'primary' } = props;
        const classes = merge(useStyles(), userClasses);

        return (
            <MUIBadge color={color} classes={classes} ref={ref} {...props} />
        );
    },
);

Badge.displayName = 'BkBadge';

export default Badge;
