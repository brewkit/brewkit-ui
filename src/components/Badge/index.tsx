import React, { ReactElement } from 'react';
import clsx from 'clsx';
import { Props } from './types';


/**
 * The Badge component is used to attach content to it's child or can be used inline with no children
 */
function Badge({
    anchorFrom = 'top-right',
    children,
    className,
    color = 'info',
    content,
    isCircular = false,
    ...otherProps
}: Props): ReactElement {


    const badgeClasses = clsx(
        'brew-Badge',
        `brew-Badge--anchorFrom-${anchorFrom}`,
        `brew-Badge--color-${color}`,
        { 'brew-Badge--isCircular': isCircular },
        { 'brew-Badge--inline': !children },
        className,
    );


    return (
        <div className={badgeClasses}>
            {children}
            <span className="brew-Badge__element" {...otherProps}>{content}</span>
        </div>
    );


}


export default Badge;
