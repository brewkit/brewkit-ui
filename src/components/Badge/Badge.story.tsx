import * as React from 'react';
import { Story } from '@storybook/react';
import MailIcon from '@material-ui/icons/Mail';
import Badge from '.';
import { BadgeProps } from '.';


export default {
    component: Badge,
    title: 'Data Display/Badge',
};

const defaultArgs = {
    invisible: false,
    showZero: false,
    max: 99,
    badgeContent: '4',
    anchorOrigin: {
        vertical: 'top' as "top",
        horizontal: 'right' as "right",
    },
    overlap: 'rectangle' as "rectangle",
    variant: 'standard' as "standard",
}
export const Sandbox: Story<BadgeProps> = (args) => <Badge {...args}><MailIcon /></Badge>;

Sandbox.args = {
    ...defaultArgs,
    color: 'primary',
};


export const Basic: Story<BadgeProps> = () => (
    <div style={{ fontSize: '2rem' }}>
        <Badge badgeContent={4} color="primary">
            ðŸ˜€
        </Badge>
        <Badge badgeContent={4} color="secondary">
            ðŸ˜€
        </Badge>
        <Badge badgeContent={4} color="error">
            ðŸ˜€
        </Badge>
    </div>
);


export const MaxValue: Story<BadgeProps> = () => (
    <div style={{ fontSize: '2rem' }}>
        <Badge badgeContent={99} color="primary">
            ðŸ˜€
        </Badge>
        <Badge badgeContent={100} color="secondary" max={99}>
            ðŸ˜€
        </Badge>
        <Badge badgeContent={1000} color="error" max={999}>
            ðŸ˜€
        </Badge>
    </div>
);


export const DotBadge: Story<BadgeProps> = () => (
    <div style={{ fontSize: '2rem' }}>

        <Badge color="primary" variant="dot">
            ðŸ˜€
        </Badge>
        <Badge color="secondary" variant="dot">
            ðŸ˜€
        </Badge>
        <Badge color="error" variant="dot">
            ðŸ˜€
        </Badge>

    </div>
);


export const Alignment: Story<BadgeProps> = () => (
    <div style={{ fontSize: '2rem' }}>

        <Badge anchorOrigin={{ vertical: 'top', horizontal: 'left' }} badgeContent={99} color="primary">
            ðŸ˜€
        </Badge>
        <Badge anchorOrigin={{ vertical: 'top', horizontal: 'right' }} badgeContent={99} color="secondary">
            ðŸ˜€
        </Badge>
        <Badge anchorOrigin={{ vertical: 'bottom', horizontal: 'left' }} badgeContent={99} color="error">
            ðŸ˜€
        </Badge>
        <Badge anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }} badgeContent={99} color="primary">
            ðŸ˜€
        </Badge>

    </div>
);
