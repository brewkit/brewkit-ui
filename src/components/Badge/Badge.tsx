import * as React from 'react';
import MuiBadge from '@material-ui/core/Badge';
import { BadgeProps } from './Badge.types';

/**
 * Badge generates a small badge to the top-right of its child(ren).
 *
 * [Material-UI Docs](https://material-ui.com/components/badges/)
 *
 * > No differences in Brewkit.
 */
export const Badge = React.forwardRef(
    (props: BadgeProps, ref: React.Ref<any>): React.ReactElement => (
        <MuiBadge ref={ref} {...props} />
    ),
);

Badge.displayName = 'Badge';

export default Badge;
