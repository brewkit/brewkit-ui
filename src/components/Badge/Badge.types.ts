import { BadgeProps } from '@material-ui/core/Badge';


export type Props = BadgeProps;
