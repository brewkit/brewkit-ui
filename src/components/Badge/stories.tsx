import React, { ReactElement } from 'react';
import { boolean, select, text } from '@storybook/addon-knobs';
import Badge from './index';
import Button from '../Button';
import Label from '../Label';
import Typography from '../Typography';


export default {
    component: Badge,
    parameters: {
        componentSubtitle: <Label color="success">In Development</Label>,
    },
    title: 'Components|Display/Badge',
};


export const general = (): ReactElement => {
    const content = text('content', '5');
    const isCircular = boolean('isCircular', false);
    const anchorFrom = select('anchorFrom', ['top-right', 'top-left', 'bottom-right', 'bottom-left'], 'top-right');
    const color = select('color', ['success', 'warning', 'danger', 'info'], 'danger');


    return (
        <Badge anchorFrom={anchorFrom} color={color} content={content} isCircular={isCircular}>
            <Label>Test</Label>
        </Badge>
    );
};


export const inline = (): ReactElement => {
    const content = text('content', 'inline');
    const isCircular = boolean('isCircular', false);
    const color = select('color', ['success', 'warning', 'danger', 'info'], 'danger');


    return (
        <div>
            <Button>
                Check it!Check it!Check it!Check it!Check it!Check it!
                <Badge content={content} color={color} isCircular={isCircular} />
            </Button>
            <br />
            <br />
            <Typography>Typography <Badge content={content} isCircular={isCircular} /></Typography>
            <br />
            <br />
            <div>Regular Div <Badge content={content} isCircular={isCircular} /></div>
        </div>
    );
};
