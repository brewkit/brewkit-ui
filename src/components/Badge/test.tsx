import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Badge from './index';


describe('Badge', () => {


    it('Renders correctly', () => {
        const tree = renderer
            .create(<Badge><div>child</div></Badge>)
            .toJSON();


        expect(tree).toMatchSnapshot();
    });


    it('Has default classes', () => {
        const badge = shallow(<Badge><div>child</div></Badge>);
        const classes = [
            '.brew-Badge',
            '.brew-Badge--anchorFrom-top-right',
            '.brew-Badge--color-info',
            '.brew-Badge__element',
        ];


        classes.forEach((className: string) => {
            expect(badge.exists(className)).toBe(true);
        });
    });


    it('Adds custom class', () => {
        const badge = shallow(<Badge className="customizer"><div>child</div></Badge>);


        expect(badge.hasClass('customizer'));
    });


    it('Has circular and inline modifier class', () => {
        const badge = shallow(<div>child <Badge isCircular /></div>);
        const classes = ['brew-Badge--isCircular', 'brew-Badge--inline'];


        classes.forEach((className: string) => {
            expect(badge.hasClass(className));
        });
    });


    it('Changes default classes to prop modifiers', () => {
        const badge = shallow(<div>child <Badge anchorFrom="bottom-right" color="danger" /></div>);
        const classes = ['brew-Badge--anchorFrom-bottom-right', 'brew-Badge--color-danger'];


        classes.forEach((className: string) => {
            expect(badge.hasClass(className));
        });
    });


});
