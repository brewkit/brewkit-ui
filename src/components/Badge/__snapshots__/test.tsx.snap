// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge Renders correctly 1`] = `
<div
  className="brew-Badge brew-Badge--anchorFrom-top-right brew-Badge--color-info"
>
  <div>
    child
  </div>
  <span
    className="brew-Badge__element"
  />
</div>
`;
